### PR TITLE
[SPARK-49537][SQL] Improve Join stats estimate 

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR improve Join stats estimate when column stats exist.

For example:
Left side:
```
36126 bytes, 2150 rows
```

info_name | info_value
-- | --
col_name | key-1980
data_type | string
comment | NULL
min | NULL
max | NULL
num_nulls | 1
distinct_count | 1980
avg_col_len | 9
max_col_len | 38
histogram | NULL

Right side:
```
13250653950 bytes, 1470064309 rows
```

info_name | info_value
-- | --
col_name | key-3896196
data_type | string
comment | NULL
min | NULL
max | NULL
num_nulls | 320713790
distinct_count | 3896196
avg_col_len | 8
max_col_len | 69
histogram | NULL

Before this PR:
```
+- Join Inner, (key-1980#388 = key-3896196#382), Statistics(sizeInBytes=71.1 MiB, rowCount=8.11E+5)
```
After this PR:
```
+- Join Inner, (key-1980#388 = key-3896196#382), Statistics(sizeInBytes=106.9 GiB, rowCount=1.25E+9)
```
The actual number of rows is [380,857,863](https://issues.apache.org/jira/secure/attachment/13071429/enable%20CBO.png). The new statistics make much more reasonable than the previous ones.

### Why are the changes needed?

1. Avoid broadcast join fail.
   ```
   org.apache.spark.SparkException: Cannot broadcast the table that is larger than 4GB: 4GB
   ```
2. Avoid Spark driver OOM.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala [hunk 1]
```diff
@@ -233,7 +233,8 @@ case class JoinEstimation(join: Join) extends Logging {
     val rightKeyStat = rightStats.attributeStats(rightKey)
     val maxNdv = leftKeyStat.distinctCount.get.max(rightKeyStat.distinctCount.get)
     // Compute cardinality by the basic formula.
-    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) / BigDecimal(maxNdv)
+    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) /
+      BigDecimal(leftKeyStat.distinctCount.get.min(rightKeyStat.distinctCount.get))
```
Comment (pavel0fadeev) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala:237]: The original formula is practically very rough, but it gives the exact cardinality if the distribution of each table is uniform and all values in the table with smaller distinctCount match in the other table.
But can you share if there is any theory behind the new formula?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala [hunk 2]
```diff
@@ -233,7 +233,8 @@ case class JoinEstimation(join: Join) extends Logging {
     val rightKeyStat = rightStats.attributeStats(rightKey)
     val maxNdv = leftKeyStat.distinctCount.get.max(rightKeyStat.distinctCount.get)
```
Comment (pavel0fadeev) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala:234]: This line is not needed after your change as I understand

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala]: Removed.

**Review by dongjoon-hyun — DISMISSED** [review]
+1, LGTM. Thank you, @wangyum .

To @cloud-fan , @yaooqinn , @viirya , @huaxingao , @sunchao , I'd like to deliver this via Apache Spark 4.0.0-preview2 next Monday. Please let me know if you have any concern on this PR.

**Comment by pavel0fadeev** [comment]
Can't write a comment to this part of the file, but there is a comment in sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala
line 161-169 that should also be aligned with this change

**Comment by dongjoon-hyun** [comment]
Thank you for the info, @pavel0fadeev .
Could you address the above comment, @wangyum ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala [hunk 3]
```diff
@@ -231,12 +231,11 @@ case class JoinEstimation(join: Join) extends Logging {
       max: Option[Any]): (BigInt, ColumnStat) = {
     val leftKeyStat = leftStats.attributeStats(leftKey)
     val rightKeyStat = rightStats.attributeStats(rightKey)
-    val maxNdv = leftKeyStat.distinctCount.get.max(rightKeyStat.distinctCount.get)
+    val newNdv = leftKeyStat.distinctCount.get.min(rightKeyStat.distinctCount.get)
     // Compute cardinality by the basic formula.
-    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) / BigDecimal(maxNdv)
+    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) / BigDecimal(newNdv)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala:236]: I don't see too much explanation why this is a better estimation in the PR description or in code comment. Could you add some?

Roughly looks like this change tries to overestimate cardinality instead underestimating previously, so it lowers the chance to broadcast join table.

Before this, if the join table fails to broadcast, user can disable broadcast to run the query. If after this change Spark doesn't broadcast the join table (as estimated states are higher than threshold), user may not be able to broadcast it (as the threshold is a hard limit).

**Comment by yaooqinn** [comment]
It looks reasonable to me to overestimate just like what we set to 'defaultSizeInBytes'

**Pull request closed without merging**